File: Shop/Controllers/CartController.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Web;
using System.Web.Mvc;
using Microsoft.AspNet.Identity;
using Newtonsoft.Json;
using Servises.Interfaces;

namespace Shop.Controllers
{
    [Authorize]
    public class CartController : Controller
    {
        private IService blService;

        public CartController(IService service)
        {
            blService = service;
        }

        [HttpGet]
        public ActionResult Cart()
        {
            return View();
        }

        [HttpGet]
        public JsonResult GetOrder()
        {
            var userName = User.Identity.GetUserName();

            var order = blService.GetOpenOrder(userName);

            var testObj = new {orderId = order.Id,
                total = order.Total,
                orderRows = order.OrderRows.Select(r => new
                {
                    rowId = r.Id,
                    productName = r.Product.Name,
                    qty = r.Qty,
                    price = r.Price,
                    sum = r.Sum
                }).ToArray()};

            return Json(testObj,JsonRequestBehavior.AllowGet);
        }

        [HttpPost]
        public ActionResult SaveOrder (string orderData)
        {

            if (orderData == null)
                return new HttpStatusCodeResult(HttpStatusCode.BadRequest);

            var data = JsonConvert.DeserializeAnonymousType(orderData, new { orderNumber = 0, comment = "", orderRows = new[] { new { rowId = 0, qty = 0m, sum = 0m} } });

            var order = blService.DatabaseService.OrderRepository.Get(o => o.Id == data.orderNumber && o.Payed == false).FirstOrDefault();

            if(order==null)
                new HttpStatusCodeResult(HttpStatusCode.NotFound);

            order.Payed = true;
            order.Comment = data.comment;
            order.Date = DateTime.Now;

            foreach(var dataRow in data.orderRows)
            {
                var orderRow = order.OrderRows.First(r => r.Id == dataRow.rowId);
                orderRow.Qty = dataRow.qty;
                orderRow.Sum = dataRow.sum;
            }

            order.Total = order.OrderRows.Sum(r => r.Sum);

            blService.DatabaseService.OrderRepository.Update(order);

            blService.DatabaseService.Save();

            return new HttpStatusCodeResult(HttpStatusCode.OK);

        }



    }
}

File: Shop/Scripts/Controllers/cartController.js
var cartController = function ($scope, $http) {
    $scope.models = {
        mode: 'cart',
        orderNumber:'',
        orderRows: undefined,
        total: 0,
        comment:''
    };

    $http.get("GetOrder").then(function(response) {

        if (response.data !== undefined) {
            $scope.models.orderRows = response.data.orderRows;
            $scope.models.total = response.data.total;
            $scope.models.orderNumber = response.data.orderId;

            if ($scope.models.orderRows.length === 0)
                $scope.models.mode = '';
        }

    });

    $scope.gotoPaymentCmd = function () {
        $scope.models.mode = 'order';

    }

    $scope.QtyChange = function (rowData) {

        rowData.sum = rowData.qty * rowData.price;

        calculateTotal();

    };

    $scope.paymentCmd = function () {

        $http.post("SaveOrder", {
            orderData: JSON.stringify({
                orderNumber: $scope.models.orderNumber,
                comment: $scope.models.comment,
                orderRows: $scope.models.orderRows
            }) }
        ).then(
            function (response) { $scope.models.mode = 'complete'; },
            function (response) { console.log(response) });


    }

    $scope.$watch('models.orderRows', calculateTotal);

    function calculateTotal () {

        var total = 0;

        if ($scope.models.orderRows !== undefined) {
            $scope.models.orderRows.forEach(function (row) { total += row.sum; })
            $scope.models.total = total;
        }

    }

}

cartController.$inject = ['$scope', '$http'];
